### ISSUE TITLE
Record names not checked with column names

### ISSUE THREAD

**Issue by vdichev** [tpolecat/doobie#277]
When mapping multi-column queries to a shapeless record, only the positions of the columns are taken into account. That is to say, the first column from the query is mapped to the first record entry, etc. This is a bit unfortunate as it's all too common to mix up the order of columns or records and if they're not the same type, then errors would be produced even though the types corresponding to the record name are correct. This is especially devious when queries select all columns via "SELECT *". Conversely, if you mix up the order and the entries are of the same type, then no error would be produced. This is hardly type-safe...

For contrast, the now abandoned sqltyped project did this the "right" way- records were constructed from the column name and type directly: https://github.com/jonifreeman/sqltyped/blob/master/core/src/main/scala/jdbc.scala

I know there are some things which can't be done at runtime (as opposed to compile-time), but on analysis (via the check method) it should be possible in doobie to check the record name with the db column name.

Is there any technical difficulty in doing this? Or is this something which is not a good design idea? After a brief look at the code, it wouldn't work using the zipped columnAlignment tuple list and it would be a bit harder to check if the number of columns is the same, but I think it would make the code safer.

**Comment by tpolecat** [comment]
Regarding `SELECT *`, this is always wrong so I'm not going to go out of my way to make it easier.

Regarding field-name checking, programs should be invariant under rename-refactoring so I don't support mapping columns to case class fields by name, for instance. There is a bit of a gray area with singleton types in shapeless records because you can make an argument that they are actually data, not code. There is no technical reason not to do this, it's just a question whether it is ethical or not.

Compile-time query checking in the style of sqltyped would make this kind of thing less of an issue, but the quality of metadata you get from the database driver is so varied (it's largely useless from MySql for example) that this is more of an "extra" when it works and not a feature to be relied upon.

In any case I will leave this open for consideration in a future version. Thanks!

**Comment by vdichev** [comment]
OK, I was suspecting that it's not only a technical issue and I see the arguments for not mapping the names to case class field names automatically.

Regarding "SELECT *"- completely agreed, it shouldn't be the main argument. I just gave this as an example since the other non-abandoned DB library which checks SQL queries- Slick- gives this example on http://slick.typesafe.com/doc/3.1.1/sql.html:

```scala
def getSuppliers(id: Int): DBIO[Seq[(Int, String, String, String, String, String)]] =
  tsql"select * from suppliers where id > $id"
```

Say what? This looks horrible and not much better than the non-typesafe alternative. 5 Strings? Which is which?

And the fact that MySQL does a rotten job of returning metadata is also not a convincing- it sounds like "if some metadata from some DBs are useless, it's useless as a whole". After all, if using this argument we could argue that analyzing via check in doobie is useless as well. I'd say it's pretty obvious, use the metadata if it's available, and punt if it's not- that sounds as valid for output return types as for column names. At any rate, using either the "generateSimpleParameterMetadata" in MySQL or alternatively connecting to a different DB like H2 with the translated schema has worked better, as described here:

http://stackoverflow.com/questions/14558353/get-the-metadata-for-prepared-statement-in-java-with-mysql-db

To summarize, I now see checking column names is not clearly the right way, but for the sake of a future discussion, not all of the arguments were convincing. It's just that having tried sqltyped, I would have thought that if you don't want to check the names you'd use something like an HList or tuples and if you do, you can use records.
